server/src/main/java/org/elasticsearch/action/ingest/SimulateIndexResponse.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.ingest;
+
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.action.support.replication.ReplicationResponse;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * This is an IndexResponse that is specifically for simulate requests. Unlike typical IndexResponses, we need to include the original
+ * source in a SimulateIndexResponse, and don't need most other fields. This has to extend IndexResponse though so that it can be used by
+ * BulkItemResponse in IngestService.
+ */
+public class SimulateIndexResponse extends IndexResponse {
+    private final BytesReference source;
+    private final XContentType sourceXContentType;
+
+    public SimulateIndexResponse(StreamInput in) throws IOException {
+        super(in);
+        this.source = in.readBytesReference();
+        this.sourceXContentType = XContentType.valueOf(in.readString());
+        setShardInfo(new ReplicationResponse.ShardInfo(0, 0));
+    }
+
+    public SimulateIndexResponse(
+        String id,
+        String index,
+        long version,
+        BytesReference source,
+        XContentType sourceXContentType,
+        List<String> pipelines
+    ) {
+        // We don't actually care about most of the IndexResponse fields:
+        super(new ShardId(index, "", 0), id == null ? "" : id, 0, 0, version, true, pipelines);
+        this.source = source;
+        this.sourceXContentType = sourceXContentType;
+        setShardInfo(new ReplicationResponse.ShardInfo(0, 0));
+    }
+
+    @Override
+    public XContentBuilder innerToXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field("_id", getId());
+        builder.field("_index", getShardId().getIndexName());
+        builder.field("_version", getVersion());
+        builder.field("_source", XContentHelper.convertToMap(source, false, sourceXContentType).v2());
+        assert executedPipelines != null; // This ought to never be null because we always ask to list pipelines in simulate mode
```

Comment:
Can you make the comment into the assertion error message? (empty assert messages are a pet peeve of mine)